src/kv.c

```diff
@@ -0,0 +1,908 @@
+    if ((ret = wolfsentry_kv_get_2(wolfsentry, &wolfsentry->user_values, key, key_len, WOLFSENTRY_KV_UINT, &kv)) < 0)
+        return ret;
+    *value = WOLFSENTRY_KV_V_UINT(&kv->kv);
+    WOLFSENTRY_RETURN_OK;
+}
+
+wolfsentry_errcode_t wolfsentry_user_value_store_sint(
+    struct wolfsentry_context *wolfsentry,
+    const char *key,
+    int key_len,
+    int64_t value,
+    int overwrite_p)
+{
+    wolfsentry_errcode_t ret;
+    struct wolfsentry_kv_pair_internal *kv;
+    if ((ret = wolfsentry_kv_new(wolfsentry, key, key_len, 0 /* data_len */, &kv)) < 0)
+        return ret;
+    WOLFSENTRY_KV_TYPE(&kv->kv) = WOLFSENTRY_KV_SINT;
+    WOLFSENTRY_KV_V_SINT(&kv->kv) = value;
+    if (overwrite_p)
+        ret = wolfsentry_kv_set(wolfsentry, &wolfsentry->user_values, kv);
+    else
+        ret = wolfsentry_kv_insert(wolfsentry, &wolfsentry->user_values, kv);
+
+    if (ret < 0)
+        WOLFSENTRY_FREE(kv);
+
+    return ret;
+}
+
+wolfsentry_errcode_t wolfsentry_user_value_get_sint(
+    struct wolfsentry_context *wolfsentry,
+    const char *key,
+    int key_len,
+    int64_t *value)
+{
+    wolfsentry_errcode_t ret;
+    struct wolfsentry_kv_pair_internal *kv = NULL;
+    if ((ret = wolfsentry_kv_get_2(wolfsentry, &wolfsentry->user_values, key, key_len, WOLFSENTRY_KV_SINT, &kv)) < 0)
+        return ret;
+    *value = WOLFSENTRY_KV_V_SINT(&kv->kv);
+    WOLFSENTRY_RETURN_OK;
+}
+
+wolfsentry_errcode_t wolfsentry_user_value_store_double(
+    struct wolfsentry_context *wolfsentry,
+    const char *key,
+    int key_len,
+    double value,
+    int overwrite_p)
+{
+    wolfsentry_errcode_t ret;
+    struct wolfsentry_kv_pair_internal *kv;
+    if ((ret = wolfsentry_kv_new(wolfsentry, key, key_len, 0 /* data_len */, &kv)) < 0)
+        return ret;
+    WOLFSENTRY_KV_TYPE(&kv->kv) = WOLFSENTRY_KV_FLOAT;
+    WOLFSENTRY_KV_V_FLOAT(&kv->kv) = value;
+    if (overwrite_p)
+        ret = wolfsentry_kv_set(wolfsentry, &wolfsentry->user_values, kv);
+    else
+        ret = wolfsentry_kv_insert(wolfsentry, &wolfsentry->user_values, kv);
+
+    if (ret < 0)
+        WOLFSENTRY_FREE(kv);
+
+    return ret;
+}
+
+wolfsentry_errcode_t wolfsentry_user_value_get_float(
+    struct wolfsentry_context *wolfsentry,
+    const char *key,
+    int key_len,
+    double *value)
+{
+    wolfsentry_errcode_t ret;
+    struct wolfsentry_kv_pair_internal *kv = NULL;
+    if ((ret = wolfsentry_kv_get_2(wolfsentry, &wolfsentry->user_values, key, key_len, WOLFSENTRY_KV_FLOAT, &kv)) < 0)
+        return ret;
+    *value = WOLFSENTRY_KV_V_FLOAT(&kv->kv);
+    WOLFSENTRY_RETURN_OK;
+}
+
+wolfsentry_errcode_t wolfsentry_user_value_store_string(
+    struct wolfsentry_context *wolfsentry,
+    const char *key,
+    int key_len,
+    const char *value,
+    int value_len /* without terminating null */,
+    int overwrite_p)
+{
+    wolfsentry_errcode_t ret;
+    struct wolfsentry_kv_pair_internal *kv;
+    if (value_len < 0)
```

Comment:
<= maybe?

Reply:
`value_len == 0` is valid -- zero-length value.

Reply:
Resolved in conversation